### ISSUE TITLE
Anomaly with union types

### ISSUE THREAD

**Issue by daminisatya** [Microsoft/TypeScript#12628]
**Code**

```ts
// CODE 1:

function add(a:number, b:number): number {
    return x + y;
}
add (1, 2);

/* ====================
CODE 2:
*/
function add(a:number, b:String | number): number {
    return x + y;
}
add(1, " "); // This shows as error in typescript. check the function parameter list.

/*=====================
CODE 3: 
*/

function add(a:number, b:any): number {
    return x + y;
}
add(1, " "); // This doesn't show any error, here any can by any data type.

```

**Expected behavior:**

The return type (of type ```number```) is accepting parameters which have type as ```number``` and ```any``` but not ```number``` and piped ```String | number```

**Actual behavior:**

The above scenario must be uniform.

**Comment by akarzazi** [comment]
The `any` type is compatible with all types.

```
function add(a:number ) {
    return a;
}
let x: any;
add(x);  // no error

// the same goes for operands:
let b: number;
b + c ; // no error

```
